### PR TITLE
[protobuf] Upgrade third_party/protobuf to 22.x

### PR DESCRIPTION
**Pull request by jtattermusch**
The very non-trivial upgrade of third_party/protobuf to 22.x
This PR strives to be as small as possible and many changes that were compatible with protobuf 21.x and didn't have to be merged atomically with the upgrade were already merged. 
Due to the complexity of the upgrade, this PR wasn't created automatically by a tool, but manually. Subsequent upgraded of third_party/protobuf with our OSS release script should work again once this change is merged.

This is best reviewed commit-by-commit, I tried to group changes in logical areas.

Notable changes:
- the upgrade of third_party/protobuf submodule, the bazel protobuf dependency itself
- upgrade of UPB dependency to 22.x  (in the past, we used to always upgrade upb to "main", but upb now has release branch as well). UPB needs to be upgraded atomically with protobuf since there's a de-facto circular dependency (new protobuf depends on new upb, which depends on new protobuf for codegen).
- some protobuf and upb bazel rules are now aliases, so ` extract_metadata_from_bazel_xml.py` and `gen_upb_api_from_bazel_xml.py` had to be modified to be able to follow aliases and reach the actual aliased targets.
- some protobuf public headers were renamed, so especially `src/compiler` needed to be updated to use the new headers.
- protobuf and upb now both depend on utf8_range project, so since we bundle upb with grpc in some languages, we now have to bundle utf8_range as well (hence changes in build for python, PHP, objC, cmake etc).
- protoc now depends on absl and utf8_range  (previously protobuf had absl dependency, but not for the codegen part), so python's make_grpcio_tools.py required partial rewrite to be able to handle those dependencies in the grpcio_tools build.
- many updates and fixes required for C++ distribtests (currently they all pass, but we'll probably need to follow up, make protobuf's and grpc's handling of dependencies more aligned and revisit the distribtests)
- bunch of other changes mostly due to overhaul of protobuf's and upb's internal build layout.

TODOs:
- [DONE] make sure IWYU and clang_tidy_code pass
- create a list of followups (e.g. work to reenable the few tests I had to disable and to remove workaround I had to use)
- [DONE in cl/523706129] figure out problem(s) with internal import

### CODE REVIEW

**Comment by jtattermusch** [comment]
@sampajano  the changes to ObjC are pretty minor, but can you please doublecheck anyway? (there's a commit to bump required target versions for objc examples and a change to add utf8_range include.)

**Comment by jtattermusch** [comment]
@gnossen  can you please review for python?
Besides modifying the build to handle the utf8_range dependency for grpc there are these changes:
- partial rewrite of make_grpcio_tools.py to be able to handle absl and utf8_range dependencies (and for better readability)
- temporarily disabling a few python bazel distribtest  (the failing tests are described in https://github.com/grpc/grpc/issues/32778) but after a lot of experimenting, it no longer feels productive to try to include the fix in this large PR, it will probably be much easier to get the upgrade PR merged and then followup with a fix, especially due to the tight deadline for the protobuf upgrade.

**Comment by jtattermusch** [comment]
@veblush after addressing the TODOs in PR description this will be ready to merge.

File: test/distrib/bazel/python/WORKSPACE [hunk 1]
```diff
@@ -20,3 +20,20 @@ local_repository(
...
+load("@upb//bazel:system_python.bzl", "system_python")
+
+system_python(
+    name = "system_python",
+    minimum_python_version = "3.7",
+)
+
+load("@system_python//:pip.bzl", "pip_parse")
+
+pip_parse(
```
Comment (gnossen) [inline test/distrib/bazel/python/WORKSPACE:33]: Are these dependencies actually necessary? I don't see why they should be. This looks like it's just pulling in numpy as a test dependency. This is also a breaking change to our downstream Python Bazel build consumers.

Comment (gnossen) [reply test/distrib/bazel/python/WORKSPACE:33]: Discussed offline. Will allow this for this PR, but this must be a blocker for the next gRPC release.

**Review by gnossen — APPROVED** [review]
Provisional LGTM given that [this issue](https://github.com/grpc/grpc/issues/32851) is resolved before the next release.

**Review by mikefan1991 — APPROVED** [review]
LGTM on objective-c

**Comment by sampajano** [comment]
> @sampajano the changes to ObjC are pretty minor, but can you please doublecheck anyway? (there's a commit to bump required target versions for objc examples and a change to add utf8_range include.)

@jtattermusch Thanks for checking. Sorry i was OOO until today. ObjC changes LGTM thanks :)

**Comment by shareefj** [comment]
Anyone got any idea what I need to do to update my flow to work with gRPC v1.55.0 and later.  I think the issue is this protobuf upgrade and I found one mention of CMake changes required so now do a `find_package(protobuf CONFIG REQUIRED)` but that hasn't helped.

The error I'm getting is an undefined reference:

```
 undefined reference to `google::protobuf::internal::ThreadSafeArena::thread_cache_'
```

and my CMakeLists.txt looks like:

```
project(genesee_protobuf LANGUAGES CXX)
cmake_minimum_required(VERSION 3.15)

find_package(protobuf CONFIG REQUIRED)
find_package(gRPC CONFIG REQUIRED)
find_package(Threads REQUIRED)

set(PROTO_FILES
    ${CMAKE_CURRENT_LIST_DIR}/genesee_protobuf/discovery/discovery.proto
    ${CMAKE_CURRENT_LIST_DIR}/genesee_protobuf/backdoor_memory_interface/backdoor_memory_interface.proto
    ${CMAKE_CURRENT_LIST_DIR}/genesee_protobuf/trace/trace.proto
)

set(PROTO_BINARY_DIR ${CMAKE_CURRENT_BINARY_DIR})
set(PROTO_IMPORT_DIR ${CMAKE_CURRENT_LIST_DIR})

add_library(genesee_protobuf ${PROTO_FILES})
target_link_libraries(genesee_protobuf PUBLIC protobuf::libprotobuf gRPC::grpc gRPC::grpc++)
target_include_directories(genesee_protobuf PUBLIC "$<BUILD_INTERFACE:${PROTO_BINARY_DIR}>")

#
# Compile protobuf and grpc files in genesee_protobuf target to cpp
#
get_target_property(grpc_cpp_plugin_location gRPC::grpc_cpp_plugin LOCATION)

protobuf_generate(
    TARGET genesee_protobuf
    IMPORT_DIRS ${PROTO_IMPORT_DIR}
    PROTOC_OUT_DIR ${PROTO_BINARY_DIR}
    LANGUAGE cpp)

protobuf_generate(
    TARGET genesee_protobuf
    LANGUAGE grpc
    IMPORT_DIRS ${PROTO_IMPORT_DIR}
    PROTOC_OUT_DIR ${PROTO_BINARY_DIR}
    GENERATE_EXTENSIONS .grpc.pb.h .grpc.pb.cc
    PLUGIN "protoc-gen-grpc=${grpc_cpp_plugin_location}")
```

My main application uses the shared library spat out by the above.

**Pull request merged**